Prevent insecure l2 interactions until chain sync

ffakenz opened the pull request:
<!-- Describe your change here -->

🔑 Motivation

> Closes [#2286](https://github.com/cardano-scaling/hydra/issues/2286)

This PR introduces new synchronization tracking (wall-clock time) and input-handling logic to improve consistency between the chain component and node state.

---

🔄 Changes

* `NodeState` has been refactored into a sum type with two variants: `NodeInSync` and `NodeCatchingUp`.

* A new policy based on the configured `ContestationPeriod` has been introduced, defining the period of time after which the node is considered out of sync with the chain. Beyond this period, the node will refuse to process new transactions or sign snapshots.

* The **head logic** now processes inputs only when the node state's current slot falls within the safe `ContestationPeriod` window relative to the latest chain time observed by the chain backend and the current wall-clock time.

  + **NodeState** transitions between `NodeInSync` and `NodeCatchingUp` according to the above policy.

  + While in `NodeInSync`, the head logic behaves as usual.

  + While in `NodeCatchingUp`:
      - **Client inputs** are **rejected** if outside the window.
      - This is the starting state.

      - **Network inputs** are **re-enqueued** (with TTL) until synchronization is restored.
         ⚠️ THIS IS PROBLEMATIC: the network component must be aware of the node state synced status and accept messages only while `NodeInSync`.

* The **API server** now notifies clients via new server outputs when the node state falls out of sync with the chain backend and when it becomes synchronized again.

---

📝 Notes

* We use the wall-clock time and not the latest known tip from the chain, as it is unreliable while the chain backend is still synchronizing with the Cardano network.

* A party member cannot go offline for longer than the configured `ContestationPeriod`; otherwise, it would be at risk, as it would be unable to contest the head closing. This violates the principle of head safety.

* We decided to move forward with this _alternative_ solution rather than the _main proposed one_ because it preserves the node architecture and allows for fast and easy testing of the protocol’s safety without requiring major hacks during end-to-end tests. This approach also makes the protocol safe by design and aware of the passage of time relative to the latest on-chain observations.
---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
draft failing e2e spec

ffakenz committed:
spike: use isSynced mutex strategy on chain backend

ffakenz committed:
spike: wait until chain is synced before starting api server

ffakenz committed:
refactor: extend Chain handle with newly introduced SyncedStatus type

> to provide the chain synced status to other components

ffakenz committed:
make use of chain sync status in api server

- to not start api server until chain is synced
- to prevent insecure actions if unsynced
- to close open ws connections if unsynced

ffakenz committed:
extend SyncedStatus with latest observed point

> to be used to compare in ws api against node-state slot

+ replaced throwing exception during ws open connection
  by sending a newly introduced ChainOutOfSync server output
+ replaced throwing exception during ws receiveInputs
  by sending InvalidInput with reason error message

ffakenz committed:
update openapi spec

ffakenz committed:
refactor synced status to compute diff based on query tip

ffakenz committed:
do not block api from start until synced

> because it prevents the node-state from catching up

ffakenz committed:
do not send a timed server output with fixed seq of 0

> as is confusing

ffakenz committed:
enhance and fix e2e scenario

ffakenz committed:
simplify synced status type

noonio commented:
Is this ready for review? It says you've requested review, but it's still in draft, and I don't see much of a description.

Happy to review when it's ready; please just adjust the statis to "Ready for Review" and add details! :) Otherwise, please don't add reviewers until it's actually ready.

ffakenz committed:
take ChainOutOfSync type out from the server output hierarchy

ffakenz committed:
head logic process inputs only when node state's current slot is within the contestation period window relative to the latest known tip

ffakenz committed:
draft logic spec to handle unsynced

ffakenz committed:
draft head logic to handle unsynced

ffakenz committed:
use fixed synced status in node spec

> to avoid failing due to unsynced

v0d1ch commented on hydra-node/src/Hydra/API/WSServer.hs:
```diff
@@ -73,13 +74,24 @@ wsApp ::
   con <- acceptRequest pending
+  -- we notify clients every time the chain is out of sync
+  -- as their inputs will get rejected
+  _ <- forkLabelled "ws-check-sync-status" $ forever $ do
```
Do you think it might be overkill to check sync status forever in a thread? Perhaps it would be better to do this on each websocket request so when the hydra-node is idle we are not doing any checks since we don't need to?

ffakenz replied:
outdated: removed

v0d1ch commented on hydra-node/src/Hydra/API/WSServer.hs:
```diff
@@ -168,8 +180,15 @@ wsApp env party tracer history callback nodeStateP networkInfoP responseChannel
       Right input -> do
-        traceWith tracer (APIInputReceived $ toJSON input)
-        callback input
+        syncedStatus <- chainSyncedStatus
```
Yes, this is what I was thinking when I wrote the comment above.

ffakenz replied:
outdated: removed

v0d1ch commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1329,19 +1327,37 @@ onClosedChainFanoutTx closedState newChainState fanoutUTxO =
+isSynced (ChainSlot nodeSlot) chainPoint Environment{contestationPeriod} =
+  let cp = fromIntegral contestationPeriod
+   in case chainPoint of
+        ChainPointAtGenesis -> 0 <= fromIntegral nodeSlot + cp
```
If we are at genesis you could just return False here right?

ffakenz replied:
outdated: removed

v0d1ch commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1329,19 +1327,37 @@ onClosedChainFanoutTx closedState newChainState fanoutUTxO =
-update env ledger NodeState{headState = st, pendingDeposits, currentSlot} ev = case (st, ev) of
+update env ledger knownTip NodeState{headState = st, pendingDeposits, currentSlot} ev = case (st, ev) of
+  -- General Unsynced
+  (_, NetworkInput{})
```
Wouldn't these pattern matches conflict with the ones below?

ffakenz replied:
outdated: removed

v0d1ch reviewed (commented):
Nice work Francositto!

ffakenz committed:
do not reject inputs in api layer

> they already rejected by the logic

ffakenz committed:
minor fix after rebase

ffakenz committed:
add comments to split logic update into sync and unsynced mode

ffakenz committed:
remove SyncedStatus type

> As it creates an unnecessary extra arrow between the node and the chain backend.
> The latest known tip will be tracked as part of Tick

ffakenz committed:
extend chain tick observation with latest known tip

ffakenz committed:
extend node state with latest known tip from tick observed

ffakenz committed:
pass current time as part of head logic update

ffakenz committed:
define unsynced policy based on cp

> use it to refine the e2e scenario

ch1bo committed:
Add NodeUnsynced and NodeSynced types

ch1bo committed:
Issue NodeUnsynced in HeadLogic when out of sync

ch1bo committed:
Draft new NodeState constructors to reflect the new behavior

ffakenz committed:
minor changes after rebase

ffakenz committed:
minor typo fix to CHANGELOG

ffakenz committed:
fix: make hydra-node compile

> this allow tests to run with failures

ffakenz committed:
fix: aggregate node state

> do not forget to update the head state using the aggregated result

ffakenz committed:
fix: logic spec

> proove we ignore network inputs when node state is catching up

ffakenz committed:
remove known tip from tick, tick observed and node state

ffakenz committed:
track current slot also while node state is catching up

ffakenz committed:
refactor: split update logic into groups that handle inputs by type

ffakenz committed:
fix: handle only chain inputs while node is catching up

ffakenz committed:
fix: handle out of sync

- to switch between synced and unsynced based on current node state

> also handle NodeSynced event during node state aggregate

ffakenz committed:
fix: golden spec files

ffakenz committed:
refactor: introduce synced status type

ffakenz committed:
wait on network inputs while node is catching up

> ignoring network messages during catch-up could result in unsigned in-flight
snapshots, as etcd would still consider them delivered.

ffakenz committed:
remove unnecessary ChainOutOfSync server output

> and fix api spec

ffakenz committed:
minor clean up of unnecessary changes

ffakenz committed:
reject client inputs instead of just ignoring them

> introduce new server output to give feedback to clients through API

ffakenz committed:
minor changes after rebase

ffakenz committed:
add node state synced status transition property

ffakenz committed:
fix: make tui to compile

ffakenz committed:
fmt: remove unused functions

ffakenz committed:
fix: e2e scenario

> by refactoring the cluster connection to node using different delay
depending on backend (blockfrost/direct)

ffakenz committed:
handle input rejected in http server as 503

ffakenz committed:
update tui to handle new server outputs and client message

ffakenz committed:
update CHANGELOG

noonio commented:
Happy to review once the build passes. Please perhaps mark as draft until such time then I can understand when to review again.

v0d1ch commented on hydra-node/src/Hydra/API/WSServer.hs:
```diff
@@ -99,7 +99,8 @@ wsApp env party tracer history callback nodeStateP networkInfoP responseChannel
   -- client.
   forwardGreetingOnly config con = do
-    NodeState{headState} <- atomically getLatestNodeState
+    nodeState <- atomically getLatestNodeState
```
I think you can revert this to reduce the diff.

ffakenz replied:
No I cant because NodeState is not a constructor anymore.

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/TimeHandle.hs:
```diff
@@ -36,6 +37,33 @@ data TimeHandleParams = TimeHandleParams
+-- * SystemStart is derived so that the given slot corresponds to the given time.
+-- * EraHistory is constructed to have a horizon sufficiently far in the future to accommodate time conversions.
+-- This is primarily used for testing purposes.
+mkTimeHandleAt :: SlotNo -> UTCTime -> TimeHandle
```
Can we move this someplace else so we know it is used in testing only?

v0d1ch reviewed (approved):
lgtm

ffakenz committed:
fix: adjust e2e assertion delays to prevent node from becoming unsynced when sending new txs

ffakenz committed:
clean up review comment
> as it is safe to use OverloadedRecordDot with sum types

ffakenz committed:
init node state as catching up

ffakenz committed:
fix e2e specs for node state now initializing in catching up

by waiting for the node to be in sync before:
- waiting for nodes to be connected
- sending init client input

ffakenz committed:
comments fix up

ffakenz committed:
tune waiting time in scenario based on how long it takes to sync

ffakenz committed:
move custom time handle factory method to logic spec

> as it is only used for testing purposes

ffakenz committed:
fix flaky e2e specs

> by waiting for node to be synced before submitting new txs

ffakenz committed:
document new synchronization polciy

ffakenz committed:
add haddock to the offline chain tick producer

ffakenz committed:
update tui to show the chain synced status

ffakenz committed:
handle node synced status in visualize logs module

ffakenz committed:
refactor tui spec to enhance scenario

vrom911 commented on hydra-node/src/Hydra/Node/State.hs:
```diff
@@ -16,14 +16,24 @@ import Test.QuickCheck (recursivelyShrink)
-  , currentSlot :: ChainSlot
-  }
+data NodeState tx
+  = -- | Normal operation of the node where it is connected and has a recent
+    -- view of the chain.
+    NodeInSync
+      { headState :: HeadState tx
+      , pendingDeposits :: PendingDeposits tx
+      -- ^ Pending deposits as observed on chain.
+      -- TODO: could even move the chain state here (also see todo below)
+      -- , chainState :: ChainStateType tx
+      , currentSlot :: ChainSlot
+      }
+  | -- | Node is catching up on its view of the chain and should behave
+    -- differently.
+    NodeCatchingUp
+      { headState :: HeadState tx
+      , pendingDeposits :: PendingDeposits tx
+      , currentSlot :: ChainSlot
+      }
```
Wouldn't that be easier to have just a field `nodeStateStatus` where `NodeStateStatus = NodeInSync | NodeCatchingUp`.

Then also we could make ToJSON/FromJSON backwards-compatible

vrom911 replied:
I see there is already `SyncedStatus` for that

ffakenz replied:
@vrom911 thanks for your feedback!

I initially modeled it the way you’re suggesting, but @ch1bo recommended changing it to the current approach during a pairing session.

Unless @ch1bo or anyone else has a new opinion on this, I’d prefer to keep it as is, otherwise, I’m happy to revert to the previous version.

vrom911 reviewed (approved):
Great work!

ffakenz committed:
minor changes after rebase

ffakenz committed:
fix another flaky e2e spec

ffakenz committed:
fix flaky e2e spec: give some time for the node to bootstrap before exception is thrown

ffakenz committed:
fix flaky e2e spec: do not wait for node synced as it reconnects fast enough and it never becomes unsynced

ffakenz committed:
fix typo: reword synch to sync

ffakenz committed:
fix flaky e2e scenario: by adding chain synced status to greetings

ffakenz committed:
enhance e2e side load spec

> by making sure the node is in sync during greetings before submitting a new tx

ffakenz committed:
enhance tui spec

ffakenz committed:
minor fix after rebase

ffakenz committed:
handle out of sync during rollback

> we included chainTime as part of Rollback chain input
> we also flipped handling logic so events get persisted
in the expected order

ch1bo commented on hydra-node/src/Hydra/Node/State.hs:
```diff
@@ -36,12 +46,23 @@ deriving anyclass instance (IsTx tx, FromJSON (ChainStateType tx)) => FromJSON (
+
+syncedStatus :: NodeState tx -> SyncedStatus
+syncedStatus NodeInSync{} = InSync
+syncedStatus NodeCatchingUp{} = OutOfSync
```
Should: make naming more consistent

We have very inconsistent naming introduced in this PR. There are at least:
- NodeInSync | NodeCatchingUp
- InSync | OutOfSync
- NodeUnsynced | NodSynced

ffakenz replied:
removed: InSync | OutOfSync

using everywhere: InSync | CatchingUp
but events are still: NodeUnsynced | NodSynced

ffakenz committed:
fix tui spec

by
- submitting an init tx before stopping the node
so on restart it does not resumes from the tip.
- making the rendering check to assert faster

ffakenz committed:
add missing update of network info projection on startup

ffakenz committed:
refactor cluster specs so that hydra-nodes wait to be in sync by default

> breaking down hydra client helpers into InSync and CatchingUp

ffakenz committed:
align synced status naming variants to improve consistency

ffakenz committed:
fix minors

- -Wmissing-fields
- -Wunused-imports
- Unnecessary comments

ffakenz committed:
update CHANGELOG

ffakenz committed:
calculate chain time for rollback, even when at genesis block

> considers the genesis block to be slot 0

ffakenz committed:
suspect node synced happens after greetings, thus racing with assertion

ffakenz committed:
fix tui specs

ffakenz committed:
simplify side load cluster spec

> also increase contestation period

noonio commented on docs/docs/configuration.md:
```diff
@@ -65,6 +65,27 @@ Depending on the upper validity picked by `hydra-node` and the current network c
+In addition to governing on-chain disputes, the contestation period is also used to determine when a Hydra node is considered **in sync** with the Cardano chain.
+
+:::important
+If a node has not observed a new block for **_half the contestation period_**, it is considered **out of sync** and transitions to a **catching up** state.
```
Can we provide an explanation of this reasoning?

Here's my thoughts:

- Imagine contestation is high ( 12 hrs ), so sync constraint is high (need to only see a block every 6 hours). This is safe because we know that we'll see two blocks within the contestation period, so ...?
- Imagine contestation is very low ( 60 seconds ) so sync requirement is very high (need to see a new block every 30 seconds); only suitable on fast netowrks ...

I.e. I'm not getting the exact scheme that determined this relationship.

noonio commented on hydra-cluster/src/HydraNode.hs:
```diff
@@ -502,16 +535,21 @@ withConnectionToNode tracer hydraNodeId =
 withConnectionToNodeHost tracer hydraNodeId apiHost@Host{hostname, port} mQueryParams action = do
   connectedOnce <- newIORef False
-  tryConnect connectedOnce (300 :: Int)
+  mBackend <- lookupEnv "HYDRA_BACKEND"
```
How come we're introducing blockfrost-style delays here?

Can I ask to also not match on simply `Just _`; if we can match on something a bit more concrete then we get a compile-time error if we add something new; not silently surprising behaviour 😅

noonio commented on hydra-cluster/test/Test/EndToEndSpec.hs:
```diff
@@ -713,6 +714,12 @@ spec = around (showLogsOnFailure "EndToEndSpec") $ do
             publishHydraScriptsAs backend Faucet
               >>= canResumeOnMemberAlreadyBootstrapped tracer tmpDir backend
 
+      it "prevent network interactions until chain backend is in sync and secure." $ \tracer -> do
```
```suggestion
      it "prevents network interactions until chain backend is in sync and secure." $ \tracer -> do
```

noonio commented on hydra-node/src/Hydra/Chain/Direct/TimeHandle.hs:
```diff
@@ -36,6 +37,10 @@ data TimeHandleParams = TimeHandleParams
 
+-- formula: 3 * k / f where k = securityParam and f = slotLength from the genesis config
+safeZone :: Pico
+safeZone = 3 * 2160 / 0.05
```
This is going to need an explanation ....

These are also network-specific, right?

noonio replied:
Oh, I see you didn't introduce this.

noonio replied:
Still, can anyone comment on the original meaning here? cc @ch1bo ?

ffakenz committed:
fix typo

Co-authored-by: Noon <noon.vandersilk@iohk.io>

noonio commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1329,74 +1327,141 @@ onClosedChainFanoutTx closedState newChainState fanoutUTxO =
+  SyncedStatus ->
+  Outcome tx
+handleOutOfSync Environment{contestationPeriod} now chainTime syncStatus
+  | chainTime `plus` CP.unsyncedPolicy contestationPeriod < now =
```
Please a comment:

"You're out of sync if last observed chainTime plus the delta allowed by the unsyncedPolicy is in the past" (or something like that)

ffakenz committed:
add comment to predicate on handleOutOfSync

ffakenz committed:
remove flaky tui spec

> because we cannot control how long the node takes to catch up. Any delay or polling on the tui test driver will be flaky

ffakenz committed:
introduced chain backend enum

> to wrap on string matching from env lookup

ffakenz committed:
enhance docs

> expand on reasoning behind: Node synchronization policy

noonio commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -213,18 +213,34 @@ findFileStartingAtDirectory maxDepth fileName = do
+    Nothing -> DefaultBackendType
+    Just "direct" -> DirectBackendType
+    Just "blockfrost" -> BlockfrostBackendType
+    Just other -> UnknownBackendType (Text.pack other)
```
This is probably just an error, right?

noonio replied:
And I guess there's no default either; in that case it's just "DirectBackendType"

Sorry, I know this is annoying given that it's only in tests; but worth not making things more confusing 😅 :)

noonio commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1329,74 +1327,144 @@ onClosedChainFanoutTx closedState newChainState fanoutUTxO =
+handleOutOfSync Environment{contestationPeriod} now chainTime syncStatus
+  -- We consider the node out of sync when:
+  -- the last observed chainTime plus the delta allowed by the unsyncedPolicy
+  -- falls behind the current system time.
```
❤️

noonio commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -1329,74 +1327,144 @@ onClosedChainFanoutTx closedState newChainState fanoutUTxO =
+    ChainInput{} ->
+      handleChainInput env ledger now currentSlot pendingDeposits st ev syncStatus
+    ClientInput{clientInput} ->
+      cause . ClientEffect $ ServerOutput.RejectedInput clientInput "chain out of sync"
```
Could we get a measure of "how" out of sync here? Something to tell us how long we might need to wait?

ffakenz replied:
We can expose "how far out of sync" the node is, but we would need to track
the current chain time explicitly. Right now the state only stores the
`currentSlot`, so we lose the information required to compute the actual drift.

If we extend the node state with a `currentChainTime :: UTCTime`, then we can always compute:

    drift = now - currentChainTime

and report that as part of the out-of-sync status.

I’m happy to add this. Are you okay with moving forward in that direction?

noonio replied:
Yes; let's do that, but let's do it in a follow up; please just make an issue and we can get the main thing here merged.

noonio commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -1280,7 +1294,13 @@ withHydraNode' dp signingKey otherParties chain action = do
       dp
-  withAsyncLabelled ("run-hydra-node", runHydraNode node) $ \_ ->
+  withAsyncLabelled ("run-hydra-node", runHydraNode node) $ \_ -> do
+    -- XXX: await for the node to be in sync with the chain
```
Why is this an "XXX" and not just a comment? Is this something we need to fix at some point? Seems like a reasonable outcome of this change?

ffakenz replied:
It is just a comment, my bad.
Removing "XXX"

noonio commented on hydra-node/test/Hydra/Model.hs:
```diff
@@ -665,9 +666,15 @@ seedWorld seedKeys seedCP futureCommits = do
-      let testClient = createTestHydraClient outputs messages outputHistory node
       nodeThread <- asyncLabelled ("seed-world-node-" <> shortLabel hsk) $ runHydraNode node
       link nodeThread
+      -- XXX: await for the node to be in sync with the chain before returning the client
```
Same comment; not sure what we need to fix here.

ffakenz replied:
It is just a comment, my bad.
Removing "XXX"

noonio commented on visualize-logs/src/VisualizeLogs.hs:
```diff
@@ -220,6 +221,8 @@ processLogs decoded =
                     LocalStateCleared{} -> pure DropLog
                     Checkpoint{} -> pure DropLog
+                    details@NodeUnsynced{} -> logIt (LogicLabel "NodeUnsynced") details
+                    details@NodeSynced{} -> logIt (LogicLabel "NodeSynced") details
```
❤️

noonio commented on CHANGELOG.md:
```diff
@@ -8,6 +8,13 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+  - Added `NodeUnsynced` and `NodeSynced` state events and server outputs.
+  - Added `RejectedInput` client message.
+  - The `Checkpoint` event, and consequently the `EventLogRotated` server output, now carry the different `NodeState` variants: `NodeInSync` or `NodeCatchingUp`.
+  - `Greetings` message now also contains the hydra-node synced status to the chain backend.
```
I think there's also something about receiving many more `Tick` messages now? Or am I understanding that wrong?

Please also link the issue :)

noonio replied:
Maybe we should also add a note that now we depend on the system wall-clock time for certain behavior; so users should take care to make sure that is set correctly.

Good to note that here, but also have a note about it in the docs (probably you already do?)

ffakenz replied:
There are no additional `Tick` messages introduced by this change. The `Tick` frequency remains exactly the same as before. What changes is only how the node interprets prolonged absence of new blocks (relative to the contestation period), not how often it receives ticks.

What made you think we were receiving many more Tick?

ffakenz replied:
We mention our dependency on wall-clock time in [docs](https://github.com/cardano-scaling/hydra/pull/2290/files#diff-26a27f31b938efc2e2e661c3ee8a5b46ea02efaf7a89e9d38d892878da24e29dR95).

noonio replied:
> What made you think we were receiving many more Tick?

In various places you added code to skip ticks, or note that a tick happened and added a + 2 somewhere.

noonio commented on hydra-node/test/Hydra/HeadLogicSpec.hs:
```diff
@@ -721,68 +763,145 @@ spec =
 
       it "ignores unrelated initTx" prop_ignoresUnrelatedOnInitTx
 
+      prop "node synced status transitions according to chain time and unsynced policy" $
```
Happy to leave it as a follow-up PR; but any chance we could get a test to check for a "bad" unsynced policy? I.e. something that we miss if we have a poor unsynced policy?

ffakenz replied:
Added a couple more:
- [safety](https://github.com/cardano-scaling/hydra/pull/2290/files#diff-0678f1fd68722b3a578009e7a011fe5a266598f7e039b434b92fc8a6d875cc32R825)
- [liveness](https://github.com/cardano-scaling/hydra/pull/2290/files#diff-0678f1fd68722b3a578009e7a011fe5a266598f7e039b434b92fc8a6d875cc32R844)

Are they enough? Is this what your where thinking about?

noonio commented on docs/docs/configuration.md:
```diff
@@ -65,6 +65,46 @@ Depending on the upper validity picked by `hydra-node` and the current network c
+    - Suitable for use-cases where blocks may be observed every large period of time.
+
+  * Low contestation ⇒ strict sync requirement
+    - Suitable for use-cases requiring fast and reliable networks where blocks are seen frequently.
```
Maybe just re-state in both of these the availability expectations; strict sync -> highly-available, node _must not_ be offline for more than half contestation.

ffakenz replied:
💯

noonio reviewed (changes requested):
Looking really good @ffakenz ; nice work.

Just a couple of points; then I think we can merge!

ffakenz committed:
fail on unknown hydra backend

ffakenz committed:
remove confusing comments marked with XXX unnecessary

ffakenz committed:
elaborate more on availability expectations

ffakenz committed:
update CHANGELOG to include a link to the issue and PR

ffakenz committed:
update CHANGELOG to mention our new dependency on 'system wall-clock time'

ffakenz committed:
add concise safety and liveness props to logic spec

noonio reviewed (approved):
Let's do it!

noonio commented on hydra-cluster/src/CardanoNode.hs:
```diff
@@ -213,18 +213,32 @@ findFileStartingAtDirectory maxDepth fileName = do
+getHydraBackend = do
+  backend <- lookupEnv "HYDRA_BACKEND"
+  pure $ case backend of
+    Nothing -> DefaultBackendType
```
Oh; please just delete this one; it's not needed (the default is "Direct").

ffakenz committed:
replace defauly hydra backend type for direct
